drivers/flash/soc_flash_b91.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2021 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT telink_b91_flash_controller
+#define FLASH_SIZE DT_REG_SIZE(DT_INST(0, soc_nv_flash))
+
+#include "flash.h"
+#include <device.h>
+#include <drivers/flash.h>
+
+
+/* driver data structure */
+struct flash_b91_data {
+	struct k_sem write_lock;
+};
+
+/* driver parameters structure */
+static const struct flash_parameters flash_b91_parameters = {
+	.write_block_size = DT_PROP(DT_INST(0, soc_nv_flash), write_block_size),
+	.erase_value = 0xff,
+};
+
+
+/* Check for correct offset and length */
+static bool flash_b91_is_range_valid(off_t offset, size_t len)
+{
+	/* check for min value */
+	if ((offset < 0) || (len < 1)) {
+		return false;
+	}
+
+	/* check for max value */
+	if ((offset + len) > FLASH_SIZE) {
+		return false;
+	}
+
+	return true;
+}
+
+/* API implementation: driver initialization */
+static int flash_b91_init(const struct device *dev)
+{
+	struct flash_b91_data *dev_data = dev->data;
+
+	k_sem_init(&dev_data->write_lock, 1, 1);
+
+	return 0;
+}
+
+/* API implementation: erase */
+static int flash_b91_erase(const struct device *dev, off_t offset, size_t len)
+{
+	int page_nums = len / PAGE_SIZE;
+	struct flash_b91_data *dev_data = dev->data;
+
+	/* check for valid length */
+	if (!len) {
```

Comment:
This condition is checked second time in `flash_b91_is_range_valid`

Reply:
Verification for `if(!len)` is not part of `flash_b91_is_range_valid` because incase of  `flash_b91_is_range_valid` we have to return `-EINVAL` error. In case of `!len` we have to return SUCCESS because it is required by `tests/drivers/flash`

Improved comment: `/* return SUCCESS if len equals 0 (required by tests/drivers/flash) */`